sign1.go

```diff
@@ -80,16 +86,22 @@ func (m *Sign1Message) UnmarshalCBOR(data []byte) error {
 		return errors.New("cbor: UnmarshalCBOR on nil Sign1Message pointer")
 	}
 
-	// fast message check
-	if !bytes.HasPrefix(data, sign1MessagePrefix) {
-		return errors.New("cbor: invalid COSE_Sign1_Tagged object")
-	}
-
 	// decode to sign1Message and parse
 	var raw sign1Message
-	if err := decModeWithTagsForbidden.Unmarshal(data[1:], &raw); err != nil {
+	var err error
+	var isUntagged bool
+
+	if bytes.HasPrefix(data, sign1MessagePrefix) {
+		err = decModeWithTagsForbidden.Unmarshal(data[1:], &raw)
+		isUntagged = false
+	} else {
+		err = decModeWithTagsForbidden.Unmarshal(data, &raw)
```

Comment:
Why? We're decoding the same thing as above (just not stripping the initial tag), and the same constraints apply. We still don't expect to see tags in the structure.

Reply:
I may be wrong, but my understanding is that `decModeWithTagsForbidden` applies only to the outer tag.

COSE does not preclude an unprotected header field to carry a tag type (e.g., a timestamp using one of the tags defined in https://datatracker.ietf.org/doc/draft-ietf-cbor-time-tag).

So, it'd be wrong if the scope of decModeWithTagsForbidden were global.

Reply:
The original  line this replaces was

```
	if err := decModeWithTagsForbidden.Unmarshal(data[1:], &raw); err != nil {
```

so `decModeWithTagsForbidden` bas being applied to the entire data starting at the second byte (byte at index 1), i.e. just stripping the initial tag. This behaviour is being preserved here.

Reply:
ok, I think we have a problem.

See https://go.dev/play/p/nXsmfeCs9cR: I have put a custom tagged time field in the unprotected header and decoding fails.  I reckon it shouldn't.

Reply:
@thomas-fossati The existing codebase is correct. Headers use `decMode` for unmarshalling. See https://github.com/veraison/go-cose/blob/3b32cdbafea6ec3443778178038c26154f7e5b7b/headers.go#L198

Reply:
@shizhMSFT I have just checked and was going to add a correcting statement :-)
Since we handle decoding in two separate steps (outer array, then headers), it works as is.